### PR TITLE
api for creating public circle form

### PR DESCRIPTION
**Pull request by zashton**

### CODE REVIEW

File: app/Http/Controllers/CircleController.php [hunk 1]
```diff
@@ -3,6 +3,7 @@
 namespace App\Http\Controllers;
```
Comment (levity) [inline app/Http/Controllers/CircleController.php:3]: is there an auto-formatter plugin for PHP, like prettier for JS? might be good to just standardize formatting around something like that.

File: routes/api.php [hunk 2]
```diff
@@ -66,10 +62,9 @@
 Route::get('/profile/{address}',[ProfileController::class, 'getProfile']);
 Route::get('/protocols', [DataController::class, 'getProtocols']);
 Route::get('/circles', [CircleController::class, 'getCircles']);
-//// not used for now
-//Route::post('/circles', [CircleController::class, 'createCircle']);
-////
-
+Route::middleware(['verify-sign-only'])->group(function() {
```
Comment (levity) [inline routes/api.php:65]: shouldn't the existing `verify-sign` work for this? because there's no `$request->route('circle_id')`

File: app/Http/Controllers/CircleController.php (hunk 1 above)
Comment (levity) [reply app/Http/Controllers/CircleController.php:3]: Not something I need you to figure out in this PR of course; just thinking out loud. do you have anything set up in your IDE that's making these formatting changes automatically?

File: routes/api.php (hunk 2 above)
Comment (zashton) [reply routes/api.php]: yeah this makes sense

File: app/Http/Controllers/CircleController.php (hunk 1 above)
Comment (zashton) [reply app/Http/Controllers/CircleController.php:3]: there is though just that it requires to trigger manually with hotkeys to autoformat , will look into that

**Comment by zashton** [comment]
adding hcaptcha as requested

**Comment by zashton** [comment]
@levity think you set protocol name as unique in the prod db when setting up retool , could i remove that restriction ?

**Comment by levity** [comment]
> @levity think you set protocol name as unique in the prod db when setting up retool , could i remove that restriction ?

Yeah, that makes sense

File: app/Http/Middleware/HCaptchaVerify.php [hunk 3]
```diff
@@ -0,0 +1,37 @@
...
+     * @param Closure $next
+     * @return mixed
+     */
+    public function handle(Request $request, Closure $next)
+    {
+        $client = new Client(['headers' => ['content-type' => 'application/x-www-form-urlencoded']]);
+        $response = $client->post('https://hcaptcha.com/siteverify',
+            [
+                'form_params' => [
+                    //h-captcha-token '10000000-aaaa-bbbb-cccc-000000000001'
```
Comment (crabsinger) [inline app/Http/Middleware/HCaptchaVerify.php:24]: should we move this commented out line?

File: app/Http/Requests/CreateCircleRequest.php [hunk 4]
```diff
@@ -0,0 +1,79 @@
...
+        }
+    }
+
+    public function rules()
+    {
+        $addressValidations = ['required', 'string', 'size:42'];
+        if ($this->protocol_id) {
+            $addressValidations[] = function ($attribute, $value, $fail) {
+                $protocol = Protocol::find($this->protocol_id);
+                if ($protocol) {
```
Comment (crabsinger) [inline app/Http/Requests/CreateCircleRequest.php:57]: Not sure this is typical in PHP and laravel, however I'm thinking that making a separate private function `is_protocol_admin(protocol_id, address)` may make it a bit more clear what this validation code is doing, and additionally allow it to be refactored more easily in the future.

File: app/Http/Requests/CreateCircleRequest.php [hunk 5]
```diff
@@ -0,0 +1,79 @@
...
+                    $circles = Circle::where('protocol_id', $this->protocol_id)->pluck('id')->toArray();
+                    if (count($circles)) {
+                        $exists = User::byAddress($value)->where('role', 1)->whereIn('circle_id', $circles)->exists();
```
Comment (crabsinger) [inline app/Http/Requests/CreateCircleRequest.php:60]: Could we make an ENUM or constant of the user roles so rather than putting `1` here, we use a constant like `USER:ROLES:ADMIN` or similar?

File: database/migrations/2021_09_27_133459_add_uxresearch_table.php [hunk 6]
```diff
@@ -0,0 +1,38 @@
...
+class AddUxresearchTable extends Migration
+{
+    /**
+     * Run the migrations.
+     *
+     * @return void
+     */
+    public function up()
+    {
+        Schema::create('uxresearch', function (Blueprint $table) {
```
Comment (levity) [inline database/migrations/2021_09_27_133459_add_uxresearch_table.php:16]: let's call this `circle_metadata` instead please (& model name "CircleMetadata")

File: app/Repositories/CircleRepository.php [hunk 7]
```diff
@@ -19,8 +23,28 @@ public function getCircles($request) {
...
+        $circle = $this->model->create(['name' => $data['circle_name'], 'protocol_id' => $protocol_id]);
+        $user = new User(['name' => $data['user_name'], 'circle_id' => $circle->id,
+            'role' => 1, 'address' => $data['address']]);
+        $user->save();
+        $profile = Profile::firstOrCreate([
+            'address' => $data['address']
+        ]);
+        $profile->load(['users.circle.protocol', 'users.teammates', 'users.histories.epoch']);
+
+        $research = new Uxresearch(['circle_id' => $circle->id, 'protocol_id' => $protocol_id,
```
Comment (levity) [inline app/Repositories/CircleRepository.php:44]: probably, if the user is already a circle admin, this data won't be present

File: app/Http/Requests/CsvRequest.php [hunk 8]
```diff
@@ -24,9 +24,9 @@ public function authorize()
     public function rules()
     {
         return [
-//            'circle_id' => 'required|integer|exists:circles,id',
```
Comment (levity) [inline app/Http/Requests/CsvRequest.php:27]: can you pull this out into its own PR please? so we can ship it sooner

**Review by levity — CHANGES_REQUESTED** [review]

File: database/migrations/2021_09_24_105720_add_circle_verification.php [hunk 9]
```diff
@@ -0,0 +1,35 @@
...
+     *
+     * @return void
+     */
+    public function up()
+    {
+        Schema::table('circles', function (Blueprint $table) {
+            $table->boolean('is_verified')->default(0);
+        });
+
+        DB::table('circles')->update(['is_verified'=> 1]);
```
Comment (crabsinger) [inline database/migrations/2021_09_24_105720_add_circle_verification.php:21]: So this will mark all old circles as `is_verified = true` after we run this migration, correct?

File: routes/api.php [hunk 10]
```diff
@@ -66,10 +62,9 @@
 Route::get('/profile/{address}',[ProfileController::class, 'getProfile']);
 Route::get('/protocols', [DataController::class, 'getProtocols']);
 Route::get('/circles', [CircleController::class, 'getCircles']);
-//// not used for now
-//Route::post('/circles', [CircleController::class, 'createCircle']);
-////
-
+Route::middleware(['hcaptcha-verify','verify-sign'])->group(function() {
+    Route::post('/create-circle', [CircleController::class, 'createCircle']);
```
Comment (crabsinger) [inline routes/api.php:66]: Typically in a RESTful API design the create for a resource `circle` would probably use a
POST to `/circles` to create a new circle.

Could we just use that RESTful pattern?


```
Common REST urls resemble the following:

    /static_root/
        GET - retrieve list of resources
            optionally include search params /static_root/?size=20&page=5
        POST - create resource
    /static_root/[id]/
        GET - retrieve one entity matching 'id'
        PUT - update entire entity
        PATCH - update partial entity
        DELETE - delete entity
   ```
https://resthooks.io/docs/guides/url

File: app/Http/Requests/CreateCircleRequest.php (hunk 5 above)
Comment (crabsinger) [reply app/Http/Requests/CreateCircleRequest.php:60]: I'm seeing this is used pretty widely so this would be good to change at some point, but not in this PR.

**Comment by crabsinger** [comment]
This PR looks good. I made a handful of comments with some ideas for improvements.

File: app/Http/Middleware/HCaptchaVerify.php (hunk 3 above)
Comment (zashton) [reply app/Http/Middleware/HCaptchaVerify.php]: removed

File: app/Http/Requests/CreateCircleRequest.php (hunk 4 above)
Comment (zashton) [reply app/Http/Requests/CreateCircleRequest.php:56]: might be premature now but will refactor once there is a repeat implementation

File: app/Http/Requests/CreateCircleRequest.php (hunk 5 above)
Comment (zashton) [reply app/Http/Requests/CreateCircleRequest.php]: updated

File: database/migrations/2021_09_27_133459_add_uxresearch_table.php [hunk 11]
```diff
@@ -0,0 +1,38 @@
...
+            $table->id();
+            $table->unsignedBigInteger('circle_id');
+            $table->unsignedBigInteger('user_id');
+            $table->unsignedBigInteger('protocol_id');
```
[the remark this answers is not in the document]
Comment (zashton) [reply database/migrations/2021_09_27_133459_add_uxresearch_table.php]: removed

File: database/migrations/2021_09_27_133459_add_uxresearch_table.php (hunk 6 above)
Comment (zashton) [reply database/migrations/2021_09_27_133459_add_uxresearch_table.php]: updated

File: app/Repositories/CircleRepository.php (hunk 7 above)
Comment (zashton) [reply app/Repositories/CircleRepository.php]: removed

File: database/migrations/2021_09_24_105720_add_circle_verification.php (hunk 9 above)
Comment (zashton) [reply database/migrations/2021_09_24_105720_add_circle_verification.php:21]: thats right

File: routes/api.php (hunk 10 above)
Comment (zashton) [reply routes/api.php]: updated

**Review by crabsinger — APPROVED** [review]
nice work on this. i think it looks good to me. just need to verify it matches FE api.

File: app/Repositories/CircleRepository.php (hunk 7 above)
Comment (levity) [reply app/Repositories/CircleRepository.php]: this is what i meant: https://github.com/coordinape/coordinape-backend/pull/47/commits/6f21ff6c553f19ffb3a89e98307017507c048939

**Comment by zashton** [comment]
@levity ah i've missed out that one, thanks for fixing that

**Comment by crabsinger** [comment]
@zashton this PR seems to be on branch staging - I think we should have it be a separate branch besides staging before merging - even though we're testing it on staging.

**Review by levity — APPROVED** [review]
